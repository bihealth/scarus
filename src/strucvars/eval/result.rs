//! Common code for evaluation results of CNVs.

use annonars::gnomad_sv::cli::query::Record as GnomadSvRecord;
use annonars::pbs::clinvar::sv::Record as ClinvarSvRecord;
use intervals_general::bound_pair::BoundPair;

use super::{del, dup};

/// Overall result.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
#[serde(rename_all = "snake_case")]
pub enum Evaluation {
    /// Result of assessing DUPs.
    Dup(dup::result::Evaluation),
    /// Result of assessing DELs.
    Del(del::result::Evaluation),
}

/// Overal assessment of a CNV.
#[derive(
    Debug,
    Clone,
    Copy,
    Default,
    PartialEq,
    strum_macros::EnumIter,
    serde::Deserialize,
    serde::Serialize,
)]
#[serde(rename_all = "snake_case")]
pub enum ClinicalSignificance {
    /// Benign
    Benign,
    /// Likely benign
    LikelyBenign,
    /// Uncertain_significance
    #[default]
    UncertainSignificance,
    /// Likely_pathogenic
    LikelyPathogenic,
    /// Pathogenic
    Pathogenic,
}

/// Score interval to use.
pub type ScoreRange = intervals_general::interval::Interval<f32>;

impl ClinicalSignificance {
    /// Return the scores,
    pub fn range(self) -> ScoreRange {
        match self {
            ClinicalSignificance::Benign => ScoreRange::UnboundedClosedRight { right: 0.99 },
            ClinicalSignificance::LikelyBenign => ScoreRange::LeftHalfOpen {
                bound_pair: BoundPair::new(-0.99, -0.90).expect("invalid interval?"),
            },
            ClinicalSignificance::UncertainSignificance => ScoreRange::Open {
                bound_pair: BoundPair::new(-0.90, 0.90).expect("invalid interval?"),
            },
            ClinicalSignificance::LikelyPathogenic => ScoreRange::LeftHalfOpen {
                bound_pair: BoundPair::new(0.90, 0.99).expect("invalid interval?"),
            },
            ClinicalSignificance::Pathogenic => ScoreRange::UnboundedClosedLeft { left: 0.99 },
        }
    }
}

/// Enumeration describing the PVS1 results.
#[derive(
    Debug,
    Clone,
    Copy,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Default,
    serde::Deserialize,
    serde::Serialize,
)]
#[serde(rename_all = "snake_case")]
pub enum Pvs1Result {
    /// PVS1
    #[default]
    Pvs1,
    /// PVS1_Strong
    Pvs1Strong,
    /// PVS1_Moderate
    Pvs1Moderate,
    /// PVS1_Supporting
    Pvs1Supporting,
}

/// Information about one overlapping ClinVar records.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct ClinvarSvOverlap {
    /// Reciprocal overlap.
    pub overlap: f32,
    /// Overlapping ClinVar record.
    pub record: ClinvarSvRecord,
}

/// Information about one overlapping gnomAD records.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct GnomadSvOverlap {
    /// Reciprocal overlap.
    pub overlap: f32,
    /// Overlapping ClinVar record.
    pub record: GnomadSvRecord,
}
