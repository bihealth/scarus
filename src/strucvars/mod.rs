//! ACMG rule prediction for structural variants.

pub mod data;
pub mod ds;
pub mod eval;

use std::path::PathBuf;

use clap::Parser;

use crate::{common::Assembly, strucvars::eval::Paths};

use self::ds::StructuralVariant;

/// Command line arguments for `strucvars` command.
#[derive(Parser, Debug)]
#[command(about = "ACMG rule applications for structural variants", long_about = None)]
pub struct Args {
    /// The used assembly.
    #[clap(long, value_enum)]
    pub assembly: Assembly,
    /// Path to Mehari transcript database file.
    #[clap(long)]
    pub path_tx_db: PathBuf,
    /// Path to HGNC xlink file.
    #[clap(long)]
    pub path_hgnc: PathBuf,
    /// Path to ClinGen dosage genes TSV file.
    #[clap(long)]
    pub path_clingen_dosage_genes: PathBuf,
    /// Path to ClinGen dosage regions TSV file.
    #[clap(long)]
    pub path_clingen_dosage_regions: PathBuf,
    /// Path to Decipher HI predictions TSV file.
    #[clap(long)]
    pub path_decipher_hi: PathBuf,
    /// Path to gnomAD constraints TSV file.
    #[clap(long)]
    pub path_gnomad_constraints: PathBuf,
    /// Path to ClinVar "minimal" RocksDB.
    #[clap(long)]
    pub path_clinvar_minimal: PathBuf,
    /// Path to functional element RocksDB.
    #[clap(long)]
    pub path_functional: PathBuf,
    /// Path to ClinVar SV RocksDB.
    #[clap(long)]
    pub path_clinvar_sv: PathBuf,
    /// Path to gnomAD SV genomes.
    #[clap(long)]
    pub path_gnomad_sv_genomes: PathBuf,
    /// Path to gnomAD SV exomes.
    #[clap(long)]
    pub path_gnomad_sv_exomes: PathBuf,

    /// The variants to prioritize.
    #[clap(required = true)]
    pub variants: Vec<StructuralVariant>,
}

/// Main entry point for the `strucvars` command.
///
/// # Arguments
///
/// * `common_args` - Commonly used command line arguments.
/// * `args` - Command line arguments specific to `strucvars` command.
///
/// # Errors
///
/// If anything goes wrong, it returns a generic `anyhow::Error`.
pub fn run(common_args: &crate::common::Args, args: &Args) -> Result<(), anyhow::Error> {
    tracing::info!("  running command `strucvars`");
    tracing::info!("  common_args = {:?}", &common_args);
    tracing::info!("  args = {:?}", &args);

    let paths = Paths {
        path_tx_db: args.path_tx_db.clone(),
        path_hgnc: args.path_hgnc.clone(),
        path_clingen_dosage_genes: args.path_clingen_dosage_genes.clone(),
        path_clingen_dosage_regions: args.path_clingen_dosage_regions.clone(),
        path_decipher_hi: args.path_decipher_hi.clone(),
        path_gnomad_constraints: args.path_gnomad_constraints.clone(),
        path_clinvar_minimal: args.path_clinvar_minimal.clone(),
        path_functional: args.path_functional.clone(),
        path_clinvar_sv: args.path_clinvar_sv.clone(),
        path_gnomad_sv_genomes: args.path_gnomad_sv_genomes.clone(),
        path_gnomad_sv_exomes: args.path_gnomad_sv_exomes.clone(),
    };
    let config = Default::default();

    let evaluator = eval::Evaluator::new(args.assembly.into(), paths, config)
        .map_err(|e| anyhow::anyhow!("failed to create evaluator: {}", e))?;
    for variant in &args.variants {
        tracing::info!("- assessing {:?}", variant);
        let result = evaluator.evaluate(variant)?;
        println!("{}", serde_json::to_string(&result)?);
    }

    Ok(())
}

#[cfg(test)]
mod test {
    #[test]
    fn run_smoke() -> Result<(), anyhow::Error> {
        let common = crate::common::Args {
            verbose: clap_verbosity_flag::Verbosity::new(1, 0),
        };

        let args = super::Args {
            assembly: crate::common::Assembly::Grch37,
            path_tx_db: "tests/data/strucvars/hi_ts/txs_example_hi.bin.zst".into(),
            path_hgnc: "tests/data/hgnc.tsv".into(),
            path_clingen_dosage_regions:
                "tests/data/strucvars/ClinGen_region_curation_list_GRCh37.tsv".into(),
            path_clingen_dosage_genes: "tests/data/strucvars/ClinGen_gene_curation_list_GRCh37.tsv"
                .into(),
            path_decipher_hi: "tests/data/strucvars/decipher_hi_prediction.tsv".into(),
            path_gnomad_constraints: "tests/data/strucvars/gnomad_constraints.tsv".into(),
            path_clinvar_minimal: "tests/data/strucvars/hi_ts/clinvar/rocksdb".into(),
            path_functional: "tests/data/strucvars/hi_ts/functional/rocksdb".into(),
            path_clinvar_sv: "tests/data/strucvars/hi_ts/clinvar-sv/rocksdb".into(),
            path_gnomad_sv_genomes: "tests/data/strucvars/hi_ts/gnomad-sv/gnomad-sv2/rocksdb"
                .into(),
            path_gnomad_sv_exomes: "tests/data/strucvars/hi_ts/gnomad-sv/exac-cnv/rocksdb".into(),
            variants: vec!["DEL:1:21000:26000".parse()?, "DUP:1:21000:26000".parse()?],
        };

        super::run(&common, &args)
    }
}
