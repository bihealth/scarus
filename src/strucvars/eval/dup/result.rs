//! Data structures for representing the actual results of DEL/copy number gain.

use crate::strucvars::{
    data::{clingen_dosage, hgnc::GeneIdInfo},
    eval::{
        common::{FunctionalElement, GeneOverlap, ScoreRange, SuggestedScore},
        result::{ClinvarSvOverlap, GnomadSvOverlap, Pvs1Result},
    },
};

/// Evaluation results for each section of the ACMG rule.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub enum Section {
    /// Results of Section G1.
    G1(G1),
    /// Results of Section G2.
    G2(G2),
    /// Results of Section G3.
    G3(G3),
    /// Results of Section G4.
    G4(G4),
}

impl SuggestedScore for Section {
    fn suggested_score(&self) -> f32 {
        match self {
            Section::G1(G1::G1A(g1a)) => g1a.suggested_score(),
            Section::G1(G1::G1B(g1b)) => g1b.suggested_score(),
            Section::G2(G2::G2A(g2a)) => g2a.suggested_score(),
            Section::G2(G2::G2B(g2b)) => g2b.suggested_score(),
            Section::G2(G2::G2C(g2c)) => g2c.suggested_score(),
            Section::G2(G2::G2D(g2d)) => g2d.suggested_score(),
            Section::G2(G2::G2E(g2e)) => g2e.suggested_score(),
            Section::G2(G2::G2F(g2f)) => g2f.suggested_score(),
            Section::G2(G2::G2G(g2g)) => g2g.suggested_score(),
            Section::G2(G2::G2H(g2h)) => g2h.suggested_score(),
            Section::G2(G2::G2I(g2i)) => g2i.suggested_score(),
            Section::G2(G2::G2J(g2j)) => g2j.suggested_score(),
            Section::G2(G2::G2L(g2l)) => g2l.suggested_score(),
            Section::G3(G3::G3A(g3a)) => g3a.suggested_score(),
            Section::G3(G3::G3B(g3b)) => g3b.suggested_score(),
            Section::G3(G3::G3C(g3c)) => g3c.suggested_score(),
            Section::G4(G4::G4Patho(g4p)) => g4p.suggested_score(),
            Section::G4(G4::G4N(g4n)) => g4n.suggested_score(),
            Section::G4(G4::G4O(g4o)) => g4o.suggested_score(),
        }
    }
}

impl ScoreRange for Section {
    fn min_score(&self) -> f32 {
        match self {
            Section::G1(G1::G1A(_)) => 0.0,
            Section::G1(G1::G1B(_)) => -0.6,
            Section::G2(G2::G2A(_)) => 1.0,
            Section::G2(G2::G2B(_)) => 0.0,
            Section::G2(G2::G2C(_)) => -1.0,
            Section::G2(G2::G2D(_)) => -1.0,
            Section::G2(G2::G2E(_)) => 0.0,
            Section::G2(G2::G2F(_)) => -1.0,
            Section::G2(G2::G2G(_)) => 0.0,
            Section::G2(G2::G2H(_)) => 0.0,
            Section::G2(G2::G2I(_)) => 0.0,
            Section::G2(G2::G2J(_)) => 0.0,
            Section::G2(G2::G2L(_)) => 0.0,
            Section::G3(G3::G3A(_)) => 0.0,
            Section::G3(G3::G3B(_)) => 0.45,
            Section::G3(G3::G3C(_)) => 0.9,
            Section::G4(G4::G4Patho(_)) => -0.9,
            Section::G4(G4::G4N(_)) => -0.9,
            Section::G4(G4::G4O(_)) => -1.0,
        }
    }

    fn max_score(&self) -> f32 {
        match self {
            Section::G1(G1::G1A(_)) => 0.0,
            Section::G1(G1::G1B(_)) => -0.6,
            Section::G2(G2::G2A(_)) => 1.0,
            Section::G2(G2::G2B(_)) => 0.0,
            Section::G2(G2::G2C(_)) => -1.0,
            Section::G2(G2::G2D(_)) => -1.0,
            Section::G2(G2::G2E(_)) => 0.0,
            Section::G2(G2::G2F(_)) => 0.0,
            Section::G2(G2::G2G(_)) => 0.0,
            Section::G2(G2::G2H(_)) => 0.0,
            Section::G2(G2::G2I(_)) => 0.9,
            Section::G2(G2::G2J(_)) => 0.0,
            Section::G2(G2::G2L(_)) => 0.0,
            Section::G3(G3::G3A(_)) => 0.0,
            Section::G3(G3::G3B(_)) => 0.45,
            Section::G3(G3::G3C(_)) => 0.9,
            Section::G4(G4::G4Patho(_)) => 0.9,
            Section::G4(G4::G4N(_)) => 0.9,
            Section::G4(G4::G4O(_)) => 0.0,
        }
    }
}

/// Enumeration of the categories for the copy number gain evaluation, Section 1.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub enum G1 {
    /// Contains protein-coding or other known functionally important elements.
    G1A(G1A),
    /// Does NOT contain protein-coding or any functionally important elements.
    G1B(G1B),
}

/// Result of the G1A subsection (important feature).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G1A {
    /// Overlapping transcripts/genes.
    pub genes: Vec<GeneOverlap>,
    /// Overlapping functional elements.
    pub functional_elements: Vec<FunctionalElement>,
}

impl SuggestedScore for G1A {
    fn suggested_score(&self) -> f32 {
        0.0
    }
}

/// Result of the G1B subsection (no important feature).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G1B {
    // no members
}

impl SuggestedScore for G1B {
    fn suggested_score(&self) -> f32 {
        -0.6
    }
}

/// Enumeration of the categories for the structural variant evaluation, Section 2.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub enum G2 {
    /// Complete overlap; Tthe TS gene or minimal criticala reegion is fully contained
    /// within the observed copy number gain.
    G2A(G2A),
    /// Partial overlap of an established TS region.
    G2B(G2B),
    /// Identical in gene content to the established benign copy number gain.
    G2C(G2C),
    /// Smaller than established benign copy number gain, breakpoint(s) does not
    /// interrupt protein-coding genes.
    G2D(G2D),
    /// Smaller than established benign copy number gain, breakpoint(s) potentially
    /// interrupt protein-coding genes.
    G2E(G2E),
    /// Larger tahn known benign copy-number gain, does not include additional
    /// protein-coding genes.
    G2F(G2F),
    /// Overlaps a benign copy number gain but includes additional material.
    G2G(G2G),
    /// TS gene fully contained within obsrved copy-number gain.
    G2H(G2H),
    /// Both breakpoints are within the asme gene (gene-level sequence variant, possibly
    /// resulting in LoF.
    G2I(G2I),
    /// One breakpoint is within an established TS gene, patient's phenotype is either
    /// inconsistent with what is expecte for LOF or that gene OR unknown
    G2J(G2J),
    /// One or both breakpoints are within gene(s) of no established clinical significance.
    G2L(G2L),
}

/// Result of the G2A subsection (complete overlap).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G2A {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping TS genes.
    pub ts_genes: Vec<clingen_dosage::Gene>,
    /// Overlapping TS genomic regions.
    pub ts_regions: Vec<clingen_dosage::Region>,
}

impl SuggestedScore for G2A {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2B subsection (partial overlap of established TS region).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G2B {
    /// Overlapping TS genomic regions.
    pub ts_regions: Vec<clingen_dosage::Region>,
}

impl SuggestedScore for G2B {
    fn suggested_score(&self) -> f32 {
        0.0
    }
}

/// Result of the G2C subsection (identical in gene content to the established
/// benign TS region).
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct G2C {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping TS genomic regions.
    pub benign_regions: Vec<clingen_dosage::Region>,
}

impl SuggestedScore for G2C {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2D subsection (smaller than established benign copy-number gain,
/// breakpoint(s) does not interrupt protein-coding gene).
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct G2D {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping benign genomic region.
    pub benign_region: clingen_dosage::Region,
}

impl SuggestedScore for G2D {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2E subsection (smaller than established benign copy-number gain,
/// breakpoint(s) potentially interrupt protein-coding gene).
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct G2E {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping benign genomic regions.
    pub benign_region: clingen_dosage::Region,
    /// Potentially affected genes.
    pub genes: Vec<GeneIdInfo>,
    /// Overlapping functional elements.
    pub functional_elements: Vec<FunctionalElement>,
}

impl SuggestedScore for G2E {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of G2F subsection (larger than known benign copy-number gain,
/// does not include additional protein-coding genes).
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct G2F {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping benign genomic regions.
    pub benign_region: clingen_dosage::Region,
}

impl SuggestedScore for G2F {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2G subsection (overlaps a benign copy-number gain but includes
/// additional genomic material).
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct G2G {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping benign genomic regions.
    pub benign_region: clingen_dosage::Region,
    /// Additional genes.
    pub genes: Vec<GeneIdInfo>,
    /// Overlapping functional elements.
    pub functional_elements: Vec<FunctionalElement>,
}

impl SuggestedScore for G2G {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2H subsection (HI gene fully contained within observed copy-number
/// gain).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G2H {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Overlapping HI gene information.
    pub hi_genes: Vec<GeneIdInfo>,
}

impl SuggestedScore for G2H {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2I subsection (both breakpoint are within the same gene, potential
/// LOF, fall back to PVS1).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G2I {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// PVS assessment result.
    pub pvs1_result: Pvs1Result,
    /// HI genes for which the assessment was made.
    pub hi_genes: Vec<GeneIdInfo>,
}

impl SuggestedScore for G2I {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2J subsection (One breakpoint is within an esbalished HI gene,
/// patient's phenotype is either inconsistent with what is expected for LOF of
/// that gene OR unknown).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G2J {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// HI genes overlapping with the CNV.
    pub hi_genes: Vec<GeneIdInfo>,
}

impl SuggestedScore for G2J {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Result of the G2L subsection (one or both breakpoints are within gene(s) of
/// no established linical significance).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G2L {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Genes overlapping with the DUP.
    pub genes: Vec<GeneIdInfo>,
}

impl SuggestedScore for G2L {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Enumeration of the categories for the structural variant evaluation, Section 2.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub enum G3 {
    /// <=34 genes.
    G3A(G3Count),
    /// 35-49 genes.
    G3B(G3Count),
    /// >=50 genes.
    G3C(G3Count),
}

/// Result of the 3A subsection (Number of protein-coding RefSeq genes wholly or partially included
/// in the copy-number loss).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G3Count {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Number of protein-coding RefSeq genes wholly or partially included in the copy-number loss.
    pub num_genes: usize,
    /// The overlapping genes.
    pub genes: Vec<GeneOverlap>,
}

impl SuggestedScore for G3Count {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}

/// Enumeration of the categories for the structural variant evaluation, Section 4.
///
/// Only 4O can be automatically determined.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub enum G4 {
    /// Overlap with pathogenic variants; must be evaluated by a human.
    ///
    /// This could be one of 4A, 4B, 4C, 4D, 4E, 4L, 4M.
    G4Patho(G4Patho),
    /// Overlap with benign variants; must be evaluated by a human.
    ///
    /// This is roughly equiavalent to 4N.
    G4N(G4N),
    /// Case–control and population evidence; Overlap with common population variation.
    G4O(G4O),
}

/// Result of the 4O subsection (overlap with common population variation).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G4Patho {
    /// Overlapping ClinVar variants, descendingly by overlap.
    pub overlaps: Vec<ClinvarSvOverlap>,
}

impl SuggestedScore for G4Patho {
    fn suggested_score(&self) -> f32 {
        0.0
    }
}

/// Benign variants from ClinVar.
///
/// This is roughly 4N.
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G4N {
    /// Overlapping ClinVar variants, descendingly by overlap.
    pub overlaps: Vec<ClinvarSvOverlap>,
}

impl SuggestedScore for G4N {
    fn suggested_score(&self) -> f32 {
        -0.9
    }
}

/// Result of the 4O subsection (overlap with common population variation).
#[derive(Debug, Clone, PartialEq, Default, serde::Deserialize, serde::Serialize)]
pub struct G4O {
    /// Suggested score for the subsection.
    pub suggested_score: f32,
    /// Accession identifiers of overlapping common variants.
    pub overlaps: Vec<GnomadSvOverlap>,
}

impl SuggestedScore for G4O {
    fn suggested_score(&self) -> f32 {
        self.suggested_score
    }
}
