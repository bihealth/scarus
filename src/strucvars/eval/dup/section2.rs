//! Implementation of evaluation of copy number gain section 2.

use bio::bio_types::genome::AbstractInterval;
use bio::bio_types::genome::Interval;

use crate::strucvars::eval::dup::result::G2C;
use crate::strucvars::{
    data::{
        clingen_dosage::{Gene, Region, Score},
        intervals::contains,
    },
    ds::StructuralVariant,
    eval::dup::result::{G2, G2B},
};

use super::result::{Section, G2A};

/// Evaluation of deletions, gain of copy number.
///
/// This is mainly used to encapsulate the functionality.  Creating new such
/// objects is very straightforward and cheap.
pub struct Evaluator<'a> {
    /// The parent evaluator.
    parent: &'a super::super::Evaluator,
}

impl<'a> Evaluator<'a> {
    /// Create a new `Evaluator`.
    pub fn with_parent(parent: &'a super::super::Evaluator) -> Self {
        Self { parent }
    }

    /// Perform the evaluation of copy number gain Section 2 and all subsection.
    ///
    /// # Arguments
    ///
    /// * `strucvar` - Structural variant to be evaluated.
    ///
    /// # Returns
    ///
    /// Returns the evaluation results of the section
    ///
    /// # Errors
    ///
    /// If anything goes wrong, it returns a generic `anyhow::Error`.
    pub fn evaluate(&self, strucvar: &StructuralVariant) -> Result<Vec<Section>, anyhow::Error> {
        tracing::debug!("evaluating section 2 for {:?}", strucvar);
        let sv_interval: Interval = strucvar.clone().into();

        // Obtain any overlapping ClinGen region and gene dosage info.
        let (clingen_genes, clingen_regions) = self.parent.clingen_overlaps(&sv_interval);

        // Check each overlapping ClinGen region/gene.  If `sv_interval` completely contains
        // a region/gene and the region/gene has a "sufficient evidence" score for TS then we
        // have case 2A.
        if let Some(result) = Self::handle_case_2a(&clingen_genes, &clingen_regions, &sv_interval) {
            tracing::debug!("case 2A fired: {:?}", &result);
            return Ok(result);
        }
        // Otherwise, we need to test case 2B.
        let has_ts_genes = clingen_genes
            .iter()
            .any(|gene| gene.triplosensitivity_score == Score::SufficientEvidence);
        let has_ts_region = clingen_regions
            .iter()
            .any(|region| region.triplosensitivity_score == Score::SufficientEvidence);
        let mut result = if has_ts_genes || has_ts_region {
            let result = vec![Section::G2(G2::G2B(G2B::default()))];
            tracing::debug!("case 2B fired: {:?}", &result);
            result
        } else {
            tracing::debug!("negative for case 2B");
            Vec::new()
        };

        // Handle case 2C and return if it fired.
        let benign_regions = clingen_regions
            .iter()
            .filter(|region| region.triplosensitivity_score == Score::DosageSensitivityUnlikely)
            .collect::<Vec<_>>();
        if let Some(result_2c) = self.handle_case_2c(&sv_interval, &benign_regions)? {
            tracing::debug!("case 2C fired: {:?}", &result_2c);
            result.push(result_2c);
            return Ok(result);
        }

        Ok(result)
    }

    /// Handle case 2A.
    fn handle_case_2a(
        clingen_genes: &[Gene],
        clingen_regions: &[Region],
        sv_interval: &bio::bio_types::genome::Interval,
    ) -> Option<Vec<Section>> {
        let ts_genes = clingen_genes
            .iter()
            .filter(|clingen_gene| {
                let clingen_interval: Interval = (*clingen_gene)
                    .clone()
                    .try_into()
                    .expect("no interval for gene");
                contains(sv_interval, &clingen_interval)
                    && clingen_gene.triplosensitivity_score == Score::SufficientEvidence
            })
            .cloned()
            .collect::<Vec<_>>();
        let ts_regions = clingen_regions
            .iter()
            .cloned()
            .filter(|clingen_region| {
                let clingen_interval: Interval = clingen_region
                    .clone()
                    .try_into()
                    .expect("no interval for region");
                contains(sv_interval, &clingen_interval)
                    && clingen_region.triplosensitivity_score == Score::SufficientEvidence
            })
            .collect::<Vec<_>>();
        if !ts_regions.is_empty() || !ts_genes.is_empty() {
            Some(vec![Section::G2(G2::G2A(G2A {
                suggested_score: 1.0,
                ts_genes,
                ts_regions,
            }))])
        } else {
            None
        }
    }

    // Handle case 2C (identical in gene content to the established benign copy-number gain).
    fn handle_case_2c(
        &self,
        sv_interval: &Interval,
        benign_regions: &[&Region],
    ) -> Result<Option<Section>, anyhow::Error> {
        // Short-circuit if there are no TS regions.
        if benign_regions.is_empty() {
            return Ok(None);
        }

        // Get sorted HGNC IDs of overlapping genes.
        let sv_hgncs = self.parent.overlapping_gene_hcnc_ids(
            sv_interval.contig(),
            sv_interval.range().start as u32,
            sv_interval.range().end as u32,
        )?;

        // Now, check each overlapping TS region for identical gene content.
        let mut res_benign_regions = Vec::new();
        for benign_region in benign_regions {
            let ts_interval: Interval = (*benign_region).clone().try_into().map_err(|e| {
                anyhow::anyhow!("failed to convert genomic location to interval: {}", e)
            })?;
            let region_hgncs = self.parent.overlapping_gene_hcnc_ids(
                ts_interval.contig(),
                ts_interval.range().start as u32,
                ts_interval.range().end as u32,
            )?;
            if sv_hgncs == region_hgncs {
                res_benign_regions.push((*benign_region).clone());
            }
        }

        if res_benign_regions.is_empty() {
            Ok(None)
        } else {
            Ok(Some(Section::G2(G2::G2C(G2C {
                suggested_score: -1.0,
                benign_regions: res_benign_regions,
            }))))
        }
    }
}

#[cfg(test)]
pub mod test {
    use hgvs::data::interface::Provider;

    use crate::strucvars::data::clingen_dosage::Score;
    use crate::strucvars::ds;

    use super::super::super::test::global_evaluator_37;

    /// Test internal working of `handle_case_2a` (complete overlap).
    #[tracing_test::traced_test]
    #[rstest::rstest]
    // Note: same cases as in `evaluate` above -- keep in sync!
    #[case("X", 123_034_319, 123_236_519, "ISCA-46743", "2A-pos")]
    #[case("X", 103_031_434, 103_047_548, "LMNB1", "2A-pos")]
    fn handle_case_2a(
        #[case] chrom: &str,
        #[case] start: u64,
        #[case] stop: u64,
        #[case] hgnc_id: &str,
        #[case] label: &str,
        global_evaluator_37: super::super::super::Evaluator,
    ) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!("{}-{}", hgnc_id, label);

        let strucvar = ds::StructuralVariant {
            chrom: chrom.into(),
            start: start as u32,
            stop: stop as u32,
            svtype: ds::SvType::Del,
            ambiguous_range: None,
        };
        let sv_interval = strucvar.into();
        let genes = global_evaluator_37
            .clingen_dosage_data
            .gene_by_overlap(&sv_interval);
        let regions = global_evaluator_37
            .clingen_dosage_data
            .region_by_overlap(&sv_interval);

        let res = super::Evaluator::handle_case_2a(&genes, &regions, &sv_interval);
        insta::assert_yaml_snapshot!(res);

        Ok(())
    }

    /// Test internal working of `handle_case_2c` (identical in gene content with benign).
    #[tracing_test::traced_test]
    #[rstest::rstest]
    // Note: same cases as in `evaluate` above -- keep in sync!
    #[case("2", 110_862_108, 110_983_703, "ISCA-37405", "2C-pos-1")] // exact match
    #[case("2", 110_802_355, 110_984_922, "ISCA-37405", "2C-pos-2")] // full: MALL, MTLN
    #[case("2", 110_862_108, 111_246_742, "ISCA-37405", "2C-neg-1")] // extra: LIMS4
    #[case("2", 110_879_569, 110_983_703, "ISCA-37405", "2C-neg-2")] // excludes: MALL
    fn handle_case_2c(
        #[case] chrom: &str,
        #[case] start: u64,
        #[case] stop: u64,
        #[case] hgnc_id: &str,
        #[case] label: &str,
        global_evaluator_37: super::super::super::Evaluator,
    ) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!("{}-{}", hgnc_id, label);
        let evaluator = super::Evaluator::with_parent(&global_evaluator_37);

        let strucvar = ds::StructuralVariant {
            chrom: chrom.into(),
            start: start as u32,
            stop: stop as u32,
            svtype: ds::SvType::Del,
            ambiguous_range: None,
        };
        let sv_interval = strucvar.into();

        let (_, clingen_regions) = global_evaluator_37.clingen_overlaps(&sv_interval);
        let benign_regions = clingen_regions
            .iter()
            .filter(|region| region.triplosensitivity_score == Score::DosageSensitivityUnlikely)
            .collect::<Vec<_>>();

        let res = evaluator.handle_case_2c(&sv_interval, &benign_regions)?;
        insta::assert_yaml_snapshot!(res);

        Ok(())
    }
}
