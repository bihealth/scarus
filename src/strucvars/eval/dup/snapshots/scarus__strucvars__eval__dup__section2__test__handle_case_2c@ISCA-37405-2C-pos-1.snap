---
source: src/strucvars/eval/dup/section2.rs
expression: res
---
G2:
  G2C:
    suggested_score: -1
    benign_regions:
      - isca_id: ISCA-37405
        isca_region_name: 2q13 recurrent region (includes NPHP1)
        genomic_location: "2:110862108-110983703"
        haploinsufficiency_score: GeneAssociatedWithRecessivePhenotype
        triplosensitivity_score: DosageSensitivityUnlikely
        haploinsufficiency_disease_id: ~
        triplosensitivity_disease_id: ~

